File: scripts/enrich_player_ratings.mjs
#!/usr/bin/env node
/** Merge data/player_ratings.csv into players_seed.json (Node fallback). */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const config = JSON.parse(fs.readFileSync(path.join(ROOT, "project.config.json"), "utf8"));
const PLAYERS_PATHS = [
  path.join(ROOT, config.seed.android, "players_seed.json"),
  path.join(ROOT, config.seed.functions, "players_seed.json"),
];
const RATINGS_CSV = path.join(ROOT, "data", "player_ratings.csv");

const strict = process.argv.includes("--strict");

function parseCsv(text) {
  const lines = text.trim().split(/\r?\n/);
  const headers = lines[0].split(",").map((h) => h.trim());
  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i];
    if (!line.trim()) continue;
    const values = [];
    let cur = "";
    let inQuotes = false;
    for (let j = 0; j < line.length; j++) {
      const ch = line[j];
      if (inQuotes) {
        if (ch === '"' && line[j + 1] === '"') {
          cur += '"';
          j++;
        } else if (ch === '"') inQuotes = false;
        else cur += ch;
      } else if (ch === '"') inQuotes = true;
      else if (ch === ",") {
        values.push(cur);
        cur = "";
      } else cur += ch;
    }
    values.push(cur);
    const row = {};
    headers.forEach((h, idx) => {
      row[h] = values[idx] ?? "";
    });
    rows.push(row);
  }
  return rows;
}

function ratingsFromRow(row) {
  const keys = [
    "overall",
    "pace",
    "shooting",
    "passing",
    "dribbling",
    "defending",
    "physical",
    "diving",
    "handling",
    "kicking",
    "reflexes",
    "speed",
    "positioning",
  ];
  const out = {};
  for (const k of keys) out[k] = parseInt(row[k], 10) || 0;
  return out;
}

function ratingsComplete(position, ratings) {
  if (ratings.overall <= 0) return false;
  if (position === "Goalkeeper") {
    return ["diving", "handling", "kicking", "reflexes", "speed", "positioning"].every((k) => ratings[k] > 0);
  }
  return ["pace", "shooting", "passing", "dribbling", "defending", "physical"].every((k) => ratings[k] > 0);
}

function main() {
  if (!fs.existsSync(RATINGS_CSV)) {
    console.error(`Missing ${RATINGS_CSV}`);
    process.exit(1);
  }
  const byId = new Map();
  for (const row of parseCsv(fs.readFileSync(RATINGS_CSV, "utf8"))) {
    const id = row.player_id?.trim();
    if (!id) continue;
    if (byId.has(id)) throw new Error(`Duplicate player_id: ${id}`);
    byId.set(id, row);
  }

  let exitCode = 0;
  for (const playersPath of PLAYERS_PATHS) {
    const players = JSON.parse(fs.readFileSync(playersPath, "utf8"));
    const enriched = players.map((p) => {
      const row = byId.get(p.playerId);
      if (!row) {
        if (strict) {
          console.error(`Missing ratings row for ${p.playerId}`);
          exitCode = 1;
        }
        return p;
      }
      const ratings = ratingsFromRow(row);
      const complete = ratingsComplete(p.position, ratings);
      if (strict && !complete) {
        console.error(`Incomplete ratings for ${p.playerId}`);
        exitCode = 1;
      }
      return {
        ...p,
        clubName: row.club_name?.trim() ?? "",
        clubLeague: row.club_league?.trim() ?? "",
        ratings,
        ratingsComplete: complete,
      };
    });
    const completeCount = enriched.filter((p) => p.ratingsComplete).length;
    fs.writeFileSync(playersPath, JSON.stringify(enriched, null, 2) + "\n", "utf8");
    console.log(`${path.relative(ROOT, playersPath)}: ${completeCount}/${enriched.length} complete`);
  }
  process.exit(exitCode);
}

main();

File: scripts/enrich_player_ratings.py
#!/usr/bin/env python3
"""
Merge data/player_ratings.csv into players_seed.json (Android + Functions).

Usage:
  python scripts/generate_seed_data.py
  python scripts/enrich_player_ratings.py
  python scripts/enrich_player_ratings.py --strict   # fail if any player incomplete
"""
from __future__ import annotations

import argparse
import csv
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
_CONFIG = json.loads((ROOT / "project.config.json").read_text(encoding="utf-8"))
PLAYERS_PATHS = [
    ROOT / _CONFIG["seed"]["android"] / "players_seed.json",
    ROOT / _CONFIG["seed"]["functions"] / "players_seed.json",
]

from player_data_lib import (  # noqa: E402
    RATINGS_CSV,
    RATINGS_CSV_COLUMNS,
    empty_ratings,
    fallback_ratings,
    load_ratings_csv,
    ratings_complete,
    ratings_from_row,
    ratings_to_csv_row,
)


def enrich_players(
    players: list[dict],
    ratings_by_id: dict[str, dict[str, str]],
    strict: bool,
    use_fallback: bool,
) -> tuple[list[dict], list[str], int]:
    errors: list[str] = []
    enriched: list[dict] = []
    fallback_count = 0

    for player in players:
        pid = player["playerId"]
        row = ratings_by_id.get(pid)
        position = player["position"]
        rarity = str(player.get("rarity", "common"))
        shirt = int(player.get("shirtNumber", 0))

        club_name = row.get("club_name", "").strip() if row else ""
        club_league = row.get("club_league", "").strip() if row else ""
        ratings = ratings_from_row(row) if row else empty_ratings()
        complete = ratings_complete(position, ratings) if row else False

        if not complete and use_fallback and not strict:
            ratings = fallback_ratings(position, rarity, shirt)
            complete = True
            fallback_count += 1
            ratings_by_id[pid] = ratings_to_csv_row(pid, ratings, club_name, club_league)
        elif row is None and strict:
            errors.append(f"Missing ratings row for player_id={pid}")
            enriched.append(player)
            continue
        elif strict and not complete:
            errors.append(f"Incomplete ratings for player_id={pid} ({player['playerName']})")
            enriched.append(player)
            continue

        player = {
            **player,
            "clubName": club_name,
            "clubLeague": club_league,
            "ratings": ratings,
            "ratingsComplete": complete,
        }
        enriched.append(player)

    extra_ids = set(ratings_by_id) - {p["playerId"] for p in players}
    for pid in sorted(extra_ids):
        errors.append(f"Ratings CSV has unknown player_id={pid}")

    return enriched, errors, fallback_count


def main() -> int:
    parser = argparse.ArgumentParser(description="Merge player_ratings.csv into players_seed.json")
    parser.add_argument("--strict", action="store_true", help="Require every player to have complete ratings")
    parser.add_argument(
        "--no-fallback",
        action="store_true",
        help="Do not synthesize stats for players missing SoFIFA data",
    )
    args = parser.parse_args()
    use_fallback = not args.no_fallback and not args.strict

    if not RATINGS_CSV.exists():
        print(f"Missing {RATINGS_CSV.relative_to(ROOT)}", file=sys.stderr)
        print("Run: python scripts/export_player_data_templates.py", file=sys.stderr)
        return 1

    ratings_by_id = load_ratings_csv()
    if not ratings_by_id:
        print(f"No rows in {RATINGS_CSV.relative_to(ROOT)}", file=sys.stderr)
        return 1

    exit_code = 0
    for path in PLAYERS_PATHS:
        if not path.exists():
            print(f"Missing {path.relative_to(ROOT)} — run generate_seed_data.py first", file=sys.stderr)
            return 1
        players = json.loads(path.read_text(encoding="utf-8"))
        enriched, errors, fallback_count = enrich_players(
            players, ratings_by_id, args.strict, use_fallback
        )
        path.write_text(json.dumps(enriched, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
        complete_count = sum(1 for p in enriched if p.get("ratingsComplete"))
        msg = f"{path.relative_to(ROOT)}: {complete_count}/{len(enriched)} players with complete ratings"
        if fallback_count:
            msg += f" ({fallback_count} fallback)"
        print(msg)
        for err in errors:
            print(f"  ! {err}", file=sys.stderr)
            exit_code = 1

    if use_fallback and ratings_by_id:
        with RATINGS_CSV.open("w", encoding="utf-8", newline="") as f:
            writer = csv.DictWriter(f, fieldnames=RATINGS_CSV_COLUMNS)
            writer.writeheader()
            for pid in sorted(ratings_by_id):
                writer.writerow(ratings_by_id[pid])
        print(f"Updated {RATINGS_CSV.relative_to(ROOT)} ({len(ratings_by_id)} rows)")

    return exit_code


if __name__ == "__main__":
    raise SystemExit(main())
